### ISSUE TITLE
Improve warning message for `fixed_regex_linter`

### ISSUE THREAD

**Issue by bersbersbers** [r-lib/lintr#1158]
> For static regular expression patterns, set `fixed = TRUE`. Note that this includes regular expressions that can be expressed as fixed patterns, e.g. [.] is really just . and \\$ is really just $ if there are no other regular expression specials. For functions from the 'stringr' package, the way to declare a static string is to wrap the pattern in stringr::fixed(). If this is being used in a dbplyr context (i.e., translated to sql), replace the regular expression with the `LIKE` operator using the `%LIKE%` infix function. Lastly, take care to remember that the `replacement` argument of `gsub()` is affected by the `fixed` argument as well.

I have a few questions/comments around  this warning, especially the second sentence.

1. "Note that this includes" - what is "this"? Why does "this" "include" something? Is this supposed to say that "the previous statement" "holds true"/"applies"/... to "regular expressions that can be expressed as fixed patterns"?
2. What does "[.]" apply to? Is that to mean that folks shouldn't use "[H]ello world", but "Hello world"? Or is this to discourage people from actually writing "[.]" and writing "." instead? I suppose it's the former, but then the use of regular expressions in statements about regular expressions is really making it unclear.
3. Why is "\\$ ... really just $ if there are no other regular expression specials"? To my understanding, `$` is an end-of-expression placeholder when `fixed = FALSE` (or omitted), and it's a literal dollar sign when `fixed = TRUE`. What do "other regular expression specials" have to do with that?
4. The part of the message in 3 feels to me like it has been added to explain *why* lintr warns about something. However, it seems to warn about everything, irrespective of any regular expression specials - e.g., expressions such as `grepl("b", "abc")`. [Is that for performance? Because I find this much more readable than `grepl("b", "abc", fixed = TRUE)`, and I am not aware of any good alternative in R.] Anyway, I wonder why lintr explains why it warns about something specific if it warns about pretty much everything.

**Comment by AshesITR** [comment]
Thanks for the feedback on the lint message. Maybe we can find a clearer way to express it.
Here's the bottom line:

1. "this" refers to regular expressions that can be expressed as a fixed (substring) expression.
2. "[.]" in regex is equivalent to "." as a fixed pattern.
3. "\$" in regex is equivalent to "$" as a fixed pattern.

Re the reasoning of linting: It is indeed performance.

```
> bench::mark(grepl("a", "abc"), grepl("a", "abc", fixed = TRUE))
# A tibble: 2 × 13
  expression                           min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result    memory time                gc                   
  <bch:expr>                      <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>    <list> <list>              <list>               
1 grepl("a", "abc")                 1.75µs   2.12µs   448246.        NA     44.8  9999     1     22.3ms <lgl [1]> <NULL> <bench_tm [10,000]> <tibble [10,000 × 3]>
2 grepl("a", "abc", fixed = TRUE)    995ns   1.26µs   734746.        NA     73.5  9999     1     13.6ms <lgl [1]> <NULL> <bench_tm [10,000]> <tibble [10,000 × 3]>
```

**Comment by MichaelChirico** [comment]
There's also a readability aspect -- regex escapes can be a lot of visual clutter:

```r
grepl("\\.\\.\\.", x)
# vs
grepl("...", x, fixed = TRUE)
```

(@AshesITR I think your point 3 should be `"\\$"` in regex, right?)

**Comment by MichaelChirico** [comment]
@bersbersbers thank you again for your intrepid usage of the bleeding-edge lintr! it's really helping us out as we push towards a major release.

the `fixed_regex_linter` wound up taking a ton of effort, so we neglected to chase down a lot of fine-tuning of the error message. I think we would do well to improve that now before release.

**Comment by MichaelChirico** [comment]
@AshesITR with the new R version of the regex, how easy would it be to split it into a few pieces to customize the message, instead of hitting it all in one giant regex?

**Comment by AshesITR** [comment]
It will be complex because you always need to match the entire regex. What could be done is test if removing one of the three token regexes still matches.
That would allow detecting if that particular token type is used.

**Comment by AshesITR** [comment]
> (@AshesITR I think your point 3 should be `"\\$"` in regex, right?)

I didn't bother to add the R escapes, since you could use raw strings to make it less painful.

**Comment by AshesITR** [comment]
How about this for a message:

> This regular expression is static, i.e., its matches can be expressed as a fixed substring expression, which is faster to compute. For static regular expression patterns, set `fixed = TRUE` or use `stringr::fixed()`.

@MichaelChirico What are the pitfalls with `gsub()` vs. `gsub(fixed = TRUE)` we need to warn about?
@bersbersbers Also keen on your feedback.

**Comment by AshesITR** [comment]
NB we could also lint `^` + static rx and static rx + `$` for the `grepl()` and `str_detect()` variants. They can be replaced by `startsWith()` and `endsWith()` respectively.

**Comment by bersbersbers** [comment]
The proposed message is perfect IMHO.

Sorry about having understood `[.]` wrong, I thought that would be equivalent to `.` *as a regex*. (Others seem to be surprised as well: https://stackoverflow.com/q/19976018/880783)

> NB we could also lint `^` + static rx and static rx + `$` for the `grepl()` and `str_detect()` variants. They can be replaced by `startsWith()` and `endsWith()` respectively, as well as always lint `^` + static rx + `$` -  it's a string equality check in disguise.

This makes a ton of sense - it's actually what I have been doing when fixing my code to make this linter happy.

An option to not warn about regexes that don't have any unnecessary escaping etc. would make sense as well. There are situations where I value readability over performance, especially when matching a single string just once.
